### PR TITLE
Sidevm as log server for cluster

### PR DESCRIPTION
**Pull request by kvinwang**

### CODE REVIEW

File: crates/pink/src/runtime/extension.rs [hunk 1]
```diff
@@ -66,6 +67,8 @@ pub fn get_side_effects() -> ExecSideEffects {
                                 error!("Contract emitted an invalid pink event");
                             }
                         }
+                    } else {
+                        result.ink_events.push((address, data));
```
Comment (h4x3rotab) [inline crates/pink/src/runtime/extension.rs:71]: Should we push link events even if `event.topics.len() != 1`?

File: pallets/phala/src/fat.rs [hunk 2]
```diff
@@ -258,6 +262,30 @@ pub mod pallet {
...
+		#[pallet::weight(0)]
+		pub fn cluster_set_log_receiver(
+			origin: OriginFor<T>,
+			cluster: ContractClusterId,
+			log_receiver: ContractId,
+		) -> DispatchResult {
+			let deployer = ensure_signed(origin)?;
+			let cluster_info = Clusters::<T>::get(&cluster).ok_or(Error::<T>::ClusterNotFound)?;
+			ensure!(
+				check_cluster_permission::<T>(&deployer, &cluster_info),
```
Comment (h4x3rotab) [inline pallets/phala/src/fat.rs:275]: `check_cluster_permission` only checks the permission for deployment. We should consider only allow the owner to set the log receiver.

File: crates/pink/src/contract.rs [hunk 3]
```diff
@@ -137,6 +139,7 @@ impl Contract {
             salt,
             block_number,
             now,
+            None,
```
Comment (h4x3rotab) [inline crates/pink/src/contract.rs:142]: No log in this case?

File: crates/pink/sidevm/sidevm/src/channel.rs [hunk 4]
```diff
@@ -81,6 +83,24 @@ impl Future for Next<'_, Message> {
...
 
+impl Future for Next<'_, SystemMessage> {
+    type Output = Option<SystemMessage>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let waker_id = crate::env::tasks::intern_waker(cx.waker().clone());
+        match ocall::poll(waker_id, self.ch.res_id.0) {
+            Ok(msg) => {
+                let message =
+                    SystemMessage::decode(&mut &msg[..]).expect("Failed to decode SystemMessage");
```
Comment (h4x3rotab) [inline crates/pink/sidevm/sidevm/src/channel.rs:94]: This way, if we introduce a new `SystemMessage`, the sidevm will crash

File: crates/pink/src/runtime/extension.rs (hunk 1 above)
Comment (kvinwang) [reply crates/pink/src/runtime/extension.rs:71]: Should

File: pallets/phala/src/fat.rs (hunk 2 above)
Comment (kvinwang) [reply pallets/phala/src/fat.rs]: Done

File: crates/pink/src/contract.rs (hunk 3 above)
Comment (kvinwang) [reply crates/pink/src/contract.rs:142]: This fn is only for unittest. no log is needed.

File: crates/pink/sidevm/sidevm/src/channel.rs (hunk 4 above)
Comment (kvinwang) [reply crates/pink/sidevm/sidevm/src/channel.rs]: Fixed

File: crates/pink/sidevm/env/src/messages.rs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        level: u8,
+        message: String,
+    },
+    PinkEvent {
+        block_number: u32,
+        contract: AccountId,
+        topics: Vec<H256>,
+        payload: Vec<u8>,
+    },
+    PinkMessageOutput {
```
[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/pink/sidevm/env/src/messages.rs:30]: There is no such thing in the pruntime that can be used to link the output to the original tx.
We may need to upgrade our mq protocol to achieve the purpose.

[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/pink/sidevm/env/src/messages.rs:30]: > I noticed the execution order of the commands are not the same as the transactions in the block

We can fix the execution order with (maybe) a little bit of effect on performance.

**Comment by h4x3rotab** [comment]
I don't have other concerns. We can have an offline discussion about the execution ordering and transaction linking later. Let's merge this PR?

**Comment by h4x3rotab** [comment]
Oh, one more thing. On our testnet, I think we will want to allow anyone to read the last N lines of the log. In this sample logger, whenever you read the log, it will be removed from the memory. So probably we should consider to make the default log just rotate, but not delete the logs so aggressively.

**Comment by kvinwang** [comment]
> Oh, one more thing. On our testnet, I think we will want to allow anyone to read the last N lines of the log. In this sample logger, whenever you read the log, it will be removed from the memory. So probably we should consider to make the default log just rotate, but not delete the logs so aggressively.

It won't remove the log after read.

**Comment by h4x3rotab** [comment]
LGTM

**Review by h4x3rotab — APPROVED** [review]

**Pull request merged**